File: java-solutions/src/google/RemoveIslands.java
package google;

import java.util.*;

/// # Remove-Islands (medium coding challenge)
/// You are given a 2d matrix consisting of 1's and 0's.
/// 1 indicates a "black" pixel and 0 indicates a "white" pixel.
/// Your task is to remove all black colored pixels that are **NOT** connected to the border
/// of the matrix, by replacing them with 0.
/// You should return the (modified) input matrix
///
/// ## Example
/// Given the following matrix
/// ```
/// let mut m1: [[u8; 6]; 6] = [
///         [1, 0, 0, 0, 0, 0],
///         [0, 1, 0, 1, 1, 1],
///         [0, 0, 1, 0, 1, 0],
///         [1, 1, 0, 0, 1, 0],
///         [1, 0, 1, 1, 0, 0],
///         [1, 0, 0, 0, 0, 1],
///     ];
/// ```
/// the cells at (1,1) (2,2), (4,2) and (4,3) should be replaced with 0
///
/// ## Output
/// return the input matrix but with the NON-connected black pixels replaced with a 0 (white) pixel


public class RemoveIslands {

    /**
     * Hold the row and col index of a "node" within a 2D matrix
     */
    private record NodeIndex(int r, int c) {
    }


    /**
     * returns NodeIndex(s) that are neighbors of the Node at index r,c in the given matrix: m
     * @param ni - r,c index of the node
     * @param m - a 2D matrix
     * @return a List of NodeIndices containing the neighbors
     */
    List<NodeIndex> neighbors(NodeIndex ni, int[][] m) {
        var ns = new ArrayList<NodeIndex>();

        // up neighbor
        if (ni.r > 0) {
            ns.add(new NodeIndex(ni.r-1, ni.c));
        }
        // down neighbor
        if (ni.r < m.length - 1) {
            ns.add(new NodeIndex(ni.r+1, ni.c));
        }
        // left neighbor
        if (ni.c > 0) {
            ns.add(new NodeIndex(ni.r, ni.c - 1));
        }
        // right neighbor
        if (ni.c < m[0].length - 1) {
            ns.add(new NodeIndex(ni.r, ni.c + 1));
        }

        return ns;
    }

    /**
     * returns true if the given NodeIndex is on the border of the matrix: m
     * @param ni
     * @param m
     * @return true if ni is a border node, else false
     */
    boolean isBorderNode(NodeIndex ni, int [][] m) {
        return ni.r == 0 || ni.r == m.length - 1 || ni.c == 0 || ni.c == m[0].length - 1;
    }

    /**
     * Iterates through all nodes of the given matrix and creates a "connectionMap" that maps
     * "1" nodes on the border, as well as their connected "1" node neighbors, to true
     * @param m
     * @return
     */
    Map<NodeIndex, Boolean> buildConnMap(int[][] m) {
        // the ConnectionMap keeps track of all nodes that are connected to a border node with value of "1"
        var cm = new HashMap<NodeIndex, Boolean>();

        for (int r = 0; r < m.length; r++) {
            for (int c = 0; c < m[0].length; c++) {
                NodeIndex curNode = new NodeIndex(r, c);

                // if curNode is a border node AND it has not yet been visited, perform a depth first search
                // to determine which other "1" nodes are connected to the border node and store them in the
                // connection map
                if (m[r][c] == 1 && isBorderNode(curNode, m) && !cm.containsKey(curNode)) {
                    Stack<NodeIndex> toVisit = new Stack<>();
                    toVisit.add(curNode);

                    while (!toVisit.isEmpty()) {
                        NodeIndex ni = toVisit.pop();
                        cm.put(ni, true);
                        List<NodeIndex> nbrs = neighbors(ni, m)
                                .stream()
                                .filter( n -> m[n.r][n.c] == 1 && !cm.containsKey(n))
                                .toList();
                        toVisit.addAll(nbrs);
                    }
                }
            }
        }
        return cm;
    }

    public static void main(String[] args) {

        int[][] m1 = new int[][]{
                {1, 0, 0, 0, 0, 0},
                {0, 1, 0, 1, 1, 1},
                {0, 0, 1, 0, 1, 0},
                {1, 1, 0, 0, 1, 0},
                {1, 0, 1, 1, 0, 0},
                {1, 0, 0, 0, 0, 1},
        };

        var ri = new RemoveIslands();

        var connMap = ri.buildConnMap(m1);

        // print the nodes that are not connected to the border
        for (int r = 0; r < m1.length; r++) {
            for (int c = 0; c < m1[0].length; c++) {
                if (m1[r][c] == 1 && !connMap.containsKey(new NodeIndex(r,c))) {
                    System.out.printf("node not connected %d:%d%n", r, c);
                    m1[r][c] = 0;
                }
            }
        }
        // print the modified matrix
        for (int[] row: m1 ) {
            System.out.println(Arrays.toString(row));
        }

    }
}

File: javascript-solutions/google/removeIslands.js
/// # Remove-Islands medium coding challenge
/// You are given a 2d matrix consisting of 1's and 0's.
/// 1 indicates a "black" pixel and 0 indicates a "white" pixel.
/// Your task is to remove all black colored pixels that are **NOT** connected to the border
/// of the matrix, by replacing them with 0.
/// You should return the (modified) input matrix
///
/// ## Example
/// Given the following matrix
/// ```
/// let mut m1: [[u8; 6]; 6] = [
///         [1, 0, 0, 0, 0, 0],
///         [0, 1, 0, 1, 1, 1],
///         [0, 0, 1, 0, 1, 0],
///         [1, 1, 0, 0, 1, 0],
///         [1, 0, 1, 1, 0, 0],
///         [1, 0, 0, 0, 0, 1],
///     ];
/// ```
/// the cells at (1,1) (2,2), (4,2) and (4,3) should be replaced with 0


/**
 * returns the indices of nodes that are adjacent to the node it index r,c. A node is adjacent if it
 * up. down, left or right
 * @param r {int}
 * @param c {int}
 * @param mtx {[[int]]} a 2d array
 * @returns [{r: int, c: int}] an array of {r, c} objects that contain the indices of neighbors nodes
 */
function neighbors(r, c, mtx) {
    const ns = [];
    // up node
    if (r > 0) {
        ns.push({ r: r-1, c: c });
    }
    // down node
    if (r < mtx.length - 1) {
        ns.push({ r: r+1, c: c });
    }
    // left node
    if (c > 0) {
        ns.push({ r: r, c: c-1 });
    }
    // right node
    if (c < mtx[0].length- 1) {
        ns.push({ r: r, c: c+1})
    }
    return ns;
}

/**
 * returns true if the given r,c index in mtx is on the border of the mtx
 * @param r {int} row index
 * @param c {int} col index
 * @param mtx {[[int]]} 2d array of integers
 * @returns {boolean}
 */
function isBorderNode(r, c, mtx) {
    return r === 0 || r === mtx.length - 1 || c === 0 || c === mtx[0].length - 1
}

/**
 * Iterates through all nodes of the given matrix and creates a "connectionMap" that maps
 * "1" nodes on the border, as well as their connected "1" node neighbors, to true
 * @param mtx {[[int]]}
 * @returns {Map<string, boolean>}
 */
function buildConnMap(mtx) {
    // the connection map, maps {r, c} object to boolean
    const cm = new Map();

    for (const [r, row] of mtx.entries()) {
        for (const [c, color] of row.entries()) {
            if (color === 1 && isBorderNode(r, c, mtx) && !cm.has(JSON.stringify({r: r, c: c}))) {
                let toVisit = [{r: r, c: c}];
                while (toVisit.length > 0) {
                    let curNode = toVisit.pop();
                    cm.set(JSON.stringify(curNode), true);
                    // get connected. non-visited neighbors of curNode
                    const nbrs = neighbors(curNode.r, curNode.c, mtx)
                        .filter(node => mtx[node.r][node.c] === 1 && !cm.has(JSON.stringify(node)));
                    toVisit =  toVisit.concat(nbrs);
                }
            }
        }
    }
    return cm;
}

const m1 = [
    [1, 0, 0, 0, 0, 0],
    [0, 1, 0, 1, 1, 1],
    [0, 0, 1, 0, 1, 0],
    [1, 1, 0, 0, 1, 0],
    [1, 0, 1, 1, 0, 0],
    [1, 0, 0, 0, 0, 1],
];

const cm = buildConnMap(m1);
for (let r = 0; r < m1.length; r++) {
    for (let c = 0; c < m1[0].length; c++) {
        if (m1[r][c] === 1 && !cm.has(JSON.stringify({r: r, c: c}))) {
            console.log(`${r}:${c} is not connected`);
        }
    }
}